src/OpenFOAM/meshes/meshShapes/face/faceFunctors.H

```diff
@@ -158,4 +159,123 @@ struct faceSweptVolFunctor
     }
 };
 
+struct faceRayFunctor
+{
+    const point* meshPoints;
+    const intersection::algorithm alg;
+    const intersection::direction dir;
+
+    faceRayFunctor
+    (
+        const point* _meshPoints,
+        const intersection::algorithm _alg,
+        const intersection::direction _dir
+    ):
+        meshPoints(_meshPoints),
+        alg(_alg),
+        dir(_dir)
+    {}
+
+    __host__ __device__
+    pointHit operator()(const faceData& tgtFace, const point& p, const vector& n)
+    {
+        const label start = tgtFace.start();
+        const label nPoints = tgtFace.size();
+
+        // If the face is a triangle, do a direct calculation
+        if (nPoints == 3)
+        {
+            return triPointRef
+            (
+                meshPoints[start],
+                meshPoints[start + 1],
+                meshPoints[start + 2]
+            ).ray(p, n, alg, dir);
+        }
+
+        // Store the points belonging to the tgtFace
+        pointField facePoints(tgtFace.size());
```

Comment:
pointField cannot be used here. It is a CPU class and cannot be used and instantiated on the GPU. This variable should be removed and all its references should be handled differently.